plenum/server/replica.py

```diff
@@ -2443,92 +704,23 @@ def _process_requested_three_phase_msg(self, msg: object,
-            return False
-        elif ((tpcKey not in self.pre_prepare_tss) or
-              ((pp, sender) not in self.pre_prepare_tss[tpcKey])):
-            return False
-        else:
-            return (abs(pp.ppTime - self.pre_prepare_tss[tpcKey][pp, sender]) <=
-                    self.config.ACCEPTABLE_DEVIATION_PREPREPARE_SECS)
-
-    def is_pre_prepare_time_acceptable(self, pp: PrePrepare, sender: str) -> bool:
-        """
-        Returns True or False depending on the whether the time in PRE-PREPARE
-        is acceptable. Can return True if time is not acceptable but sufficient
-        PREPAREs are found to support the PRE-PREPARE
-        :param pp:
-        :return:
-        """
-        key = (pp.viewNo, pp.ppSeqNo)
-        if key in self.requested_pre_prepares:
-            # Special case for requested PrePrepares
-            return True
-        correct = self.is_pre_prepare_time_correct(pp, sender)
-        if not correct:
-            if key in self.pre_prepares_stashed_for_incorrect_time and \
-                    self.pre_prepares_stashed_for_incorrect_time[key][-1]:
-                self.logger.debug('{} marking time as correct for {}'.format(self, pp))
-                correct = True
-            else:
-                self.logger.warning('{} found {} to have incorrect time.'.format(self, pp))
-        return correct
-
-    def _process_stashed_pre_prepare_for_time_if_possible(
-            self, key: Tuple[int, int]):
-        """
-        Check if any PRE-PREPAREs that were stashed since their time was not
-        acceptable, can now be accepted since enough PREPAREs are received
-        """
-        self.logger.debug('{} going to process stashed PRE-PREPAREs with '
-                          'incorrect times'.format(self))
-        q = self.quorums.f
-        if len(self.preparesWaitingForPrePrepare[key]) > q:
-            times = [pr.ppTime for (pr, _) in
-                     self.preparesWaitingForPrePrepare[key]]
-            most_common_time, freq = mostCommonElement(times)
-            if self.quorums.timestamp.is_reached(freq):
-                self.logger.debug('{} found sufficient PREPAREs for the '
-                                  'PRE-PREPARE{}'.format(self, key))
-                stashed_pp = self.pre_prepares_stashed_for_incorrect_time
-                pp, sender, done = stashed_pp[key]
-                if done:
-                    self.logger.debug('{} already processed PRE-PREPARE{}'.format(self, key))
-                    return True
-                # True is set since that will indicate to `is_pre_prepare_time_acceptable`
-                # that sufficient PREPAREs are received
-                stashed_pp[key] = (pp, sender, True)
-                self.process_three_phase_msg(pp, sender)
-                return True
-        return False
-
-    def send(self, msg, stat=None) -> None:
+    def send(self, msg, to_nodes=None) -> None:
```

Comment:
Do we use it with `to_nodes` not equal to None?

Reply:
I'm not sure. Looks like this parameter is not used anywhere. I will remove it in the next commit.

Reply:
Checked it again. We use  ```to_nodes``` to send MessageReq for preprepare only to primary